### ISSUE TITLE
Support for `default` HTTP response

### ISSUE THREAD

**Issue by metasim** [twilio/guardrail#96]
According to the [OpenAPI Spec 2.0](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md) the `responses` section can declare a `default` response.

```yaml
      responses:
        '200':
          description: Good things happened
        default:
          description: An error occurred.
          schema:
            $ref: '#/definitions/exception'
...
definitions:
  exception:
    type: object
    required:
      - code
    properties:
      code:
        type: string
      description:
        type: string
```

When running codegen from sbt the following error is reported:

```
Unknown HTTP type: default
```

**Comment by blast-hardcheese** [comment]
Duplicate of #84

This would be straightforward to implement for the server generator, though the client will need some rework. Currently, only the "best" success response is used, which works for most APIs I try to consume, everything else is not considered for decoding (even if the error mapping is specified).

I'm away from my laptop for the next few weeks; if this is a blocker, I'm available to suggest implementation strategies.

**Comment by beatkyo** [comment]
Hum, not a #84 duplicate for me.

This one is for `default` HTTP response. Meaning, any response code not explicitly declared otherwise. This is useful to enforce an error model throughout  an entire API without the need to state every response code over and over.

**Comment by blast-hardcheese** [comment]
Ah! My mistake, I was trying to quickly respond on mobile, but I linked an unrelated issue; I thought we had a ticket for the default response already.

**Comment by tomasherman** [comment]
hey, any progress on this? it is quite important for us to get this working ... we would be willing to help if you provided some guidance

**Comment by blast-hardcheese** [comment]
@tomasherman Both the server and client have full support for status codes with different objects, which wasn't the case when this ticket was opened.

Personally, I don't have a use for `default`, so I'm not sure what the usage would look like (for both clients and servers).

If you could provide some samples of how you'd prefer the interface to be used (in the case of `default` responses), I'd be interested in taking a look!

**Comment by tomasherman** [comment]
Well, i will talk about it with guys at work on monday, but i think what we would like is to simply have another response type that can be returned, much like `Ok`, `BadRequest` or `InternalServerError` there would be `Default` ... only difference would be it would contain actual status code returned

**Comment by blast-hardcheese** [comment]
This is going to need to be a very compelling argument; `default: #/definitions/ErrorCode` would mean `ErrorCode` would be a valid response to `100 Continue`, for instance. I'm extremely wary of adding support for this use case.

One of the design decisions was to return the underlying router object from the framework itself (in the case of akka-http, a `Route` object, for example).

This permits composition with whatever domain modeling you need to do outside of the well-specc'd environment.

The downside is that you'd need to replicate the routing structure in your underlying framework, which reduces the value of codegen in the first place.

In our specifications, we provide explicit mappings for all 4xx and 5xx level errors we care to support, which results in the arguments to `fold` being nearly identical for similarly handled use cases.

For us, this is a win, as it is very obvious that the response should be different for `HTTP 407 Proxy Authentication Required` vs `413 Request Entity Too Large`

Hopefully this provides some context as to why `default` is not supported directly.

The architecture of guardrail is intended to be heavily extensible, via cats' `FunctionK` composition, so even if there isn't a compelling and safe way to add the functionality to guardrail directly, you will likely only need to provide overrides for a few functions in order to provide the functionality yourself in an internally wrapped version (as we've done, for Twilio-centric idioms and standard types).

**Comment by tomasherman** [comment]
I completely get your argument and agree that errors should be 'typed' and using default is antipattern imho. Unfortunately people disagree with me and use them anyway. 

So do i get it right, that guardrail is opinionated in the functionality it provides? The reality is that people do use `default` and if guardrail aims to be universal codegen, it should probably support it

**Comment by blast-hardcheese** [comment]
@tomasherman I haven't seen a good argument for adding it. For clients, you already get the raw response in the `Left`, so you can handle it how you want there. For servers, you know what error codes you will respond with. Additionally, `x-server-raw-response` gives full control back to the user if necessary.

I'll say again, I'm still interested in hearing a solid argument as to why the quality of the API descriptions we use should degrade due to convenience. I'm all for increasing convenience (that's actually the intent of guardrail) but we can't lose quality in the process.

**Comment by tomasherman** [comment]
I don't have a better argument than 'swagger supports it' ... Would it be possible to at least make guardrail ignore it without failing? Maybe that would be enough to make guardrail more compatible with what's out there

**Comment by blast-hardcheese** [comment]
That would definitely be a reasonable compromise.

**Comment by blast-hardcheese** [comment]
I wonder if `default` in the client could be reflected as `Left((httpResponse, entity))`, though if decoding fails that needs to be reflected as well

**Comment by tomasherman** [comment]
Might be, i still think that generating additional case would be best and least confusing ... but i agree that it would enabled people less specified schmas ....

btw just some background of where i'm coming from ... we are currently having bunch of services created by bunch of teams ... and unfortunately we can't enforce any constraints on swagger schemas on other teams so whatever they decide to use in the swagger, we have to go with that ... and we simply take the json they provide and use that ... we can't even modify it :( that is why i am strongly in favour the swagger tools to allow for all functionality that swagger provides

we are really happy this project exists, we were on the brink of writing our own tool

**Comment by blast-hardcheese** [comment]
Yeah; we're actually in the same position, but this is why I'm taking such a hard line against adding `default` support to the _server_. As folks onboard onto principled tooling, it will guide people towards explicitness and away from warts like `default`, but for consumers there's much less flexibility.

For clients, this is already supported, as I mentioned, in `Left` of the result, so the minimal change would be turning the error into a warning.

**Comment by kelnos** [comment]
What I'd suggest:

For servers, do nothing.  Emit a warning when we see a `default` response.  Agreed that catch-alls in a server is a smell.

For clients, it'd be nice if everyone could be explicit, but it appears that's might be too much to ask.  Honestly, it's not bad how it is now (unknown response codes returning an error containing the raw framework response object), but we could maybe do it a little nicer.  Since we're likely already emitting the definition and decoder for the default response type, we could provide an extra method on the error type in the `Left` that unmarshals the response to it.  At the very least, we need to instead  a warning and not bail on generation.

**Comment by dsilvasc** [comment]
One use case for servers is migration from Jersey to Guardrail.  If you use the annotations documented at https://github.com/swagger-api/swagger-core/wiki/Swagger-Core-Jersey-2.X-Project-Setup-1.5 on a Jersey server to generate a swagger.yml for it, the responses will all be marked as default instead of 200.

**Comment by blast-hardcheese** [comment]
@dsilvasc That seems like a bug in the swagger generator. If they can't adequately model the different specifications of the different response codes, the generated specification won't be accurate anyway. Better to use the generated spec as a starting point, or to alter their annotations to permit explicitly specifying schemas for different response codes. I think the best thing we can do currently is to ignore the `default` for server generations and attempt to map `default` to a fallback schema in the case of unknown status codes, falling back to the raw response in the `Left`.

**Comment by dsilvasc** [comment]
@blast-hardcheese That seems reasonable.  Right now I take the generated swagger and replace all defaults with 200s before feeding it to guardrail.
